x-pack/metricbeat/module/oracle/connection.go

```diff
@@ -37,28 +31,30 @@ func init() {
 
 // NewConnection returns a connection already established with Oracle
 func NewConnection(c *ConnectionDetails) (*sql.DB, error) {
-	params, err := godror.ParseConnString(c.Hosts[0])
+	params, err := godror.ParseDSN(c.Hosts[0])
 	if err != nil {
-		return nil, errors.Wrap(err, "error trying to parse connection string in field 'hosts'")
+		return nil, fmt.Errorf("error trying to parse connection string in field 'hosts': %w", err)
 	}
 
+	// If username and password are given in separate fields in the configuration then use them to authenticate
 	if params.Username == "" {
 		params.Username = c.Username
 	}
 
-	if params.Password == "" {
-		params.Password = c.Password
+	if params.Password.Secret() == "" {
+		params.StandaloneConnection = true
+		params.Password = dsn.NewPassword(c.Password)
```

Comment:
Would it be possible to move all this url parsing logic to an url parser? This would be a more natural way to set later the `service.name` for example.

Look for example to the url parser for postgresql: https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/postgresql.go#L55

I think that the host set [here](https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/postgresql.go#L98) is later used as `service.address`.

Reply:
Thanks for passing the references. However, this logic is not for URL parsing but is basically for using the `username` and `password` if they are mentioned as the separate fields in the configuration. 

Please refer to the last two fields (commented) in the below configuration:
```
- module: oracle
  period: 10s
  metricsets:
    - performance
  enabled: true
  hosts: ['user="user" password="pass" connectString="0.0.0.0:1521/ORCLPDB1.localdomain"']

  # username: ""
  # password: ""
```

This is the reason why we haven't moved to a URL parser. Also, it would probably make it a bit more complex.
What are your thoughts?

Reply:
> However, this logic is not for URL parsing but is basically for using the `username` and `password` if they are mentioned as the separate fields in the configuration.

Well, it does a bit of both things, it parses the URL, and configures `username` and `password` if provided. Then it generates another URL, stored in `URI`, that can be used by the metricset. Like [here](https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/metricset.go#L47) in the postgresql case.

> This is the reason why we haven't moved to a URL parser. Also, it would probably make it a bit more complex.
What are your thoughts?

Not sure if more complex, but using an URL parser has some advantages:
* It is a single place to handle particularities of connection strings for certain services.
* When parsing the configured host, it allows to set in the `HostData` the `Host`, that is later used for `service.address`.
* It abstracts the handling of multiple `hosts`, each metricset instance receives a single `HostData` with the host they have to use.
* Apart of `username` and `password`, it could handle other module-specific settings related to the url/connection string.

I think that using `HostData` is the only way a metricset instance has to know what host it should use, when multiple `hosts` are configured. And the URL parser is the responsible of configuring this `HostData`.

Reply:
Understood, thanks for explaining the advantages! 
I believe we would need to refactor the module as per https://github.com/elastic/beats/issues/31611 in order to move forward with this PR.